src/palace/manager/celery/tasks/work.py

```diff
@@ -44,16 +43,24 @@ def calculate_work_presentations(
                     ),
                 ):
                     for wp in work_policies:
-                        work = get_one(session, Work, id=wp.work_id)
-                        if not work:
-                            task.log.warning(
-                                f"No work with id={wp.work_id}. Skipping..."
-                            )
-                            continue
-                        work.calculate_presentation(policy=wp.policy)
+                        with task.transaction() as tx:
+                            work = get_one(tx, Work, id=wp.work_id)
+                            if not work:
+                                task.log.warning(
+                                    f"No work with id={wp.work_id}. Skipping..."
+                                )
+                                continue
+                            work.calculate_presentation(policy=wp.policy)
+                            remaining_work_policies.remove(wp)
             except Exception as e:
                 # if a failure occurs requeue the items so that can be recalculated in the next round
-                waiting.add(*work_policies)
+                task.log.error(
+                    f"Failed to calculate_presentation for {wp.work_id}. "
+                    f"Re-queuing remaining {len(remaining_work_policies)} of {len(work_policies)} "
+                    f"for future processing.",
+                    exc_info=e,
+                )
+                waiting.add(*remaining_work_policies)
```

Comment:
Don't have to deal with this in this PR, but I am a bit worried with the structure here that an exception that isn't easily handled by a retry will just cause an infinite loop.

Like if something in `calculate_presentation` raised a `ValueError`. That same work would get added back into the set, and just raise a `ValueError` again next time the task runs. If we're going to re-add tasks to the set like this, we should probably fail them after a certain number of retries.

With the structure here, a retry count is a bit hard to implement, but if we process one work per task, it would be fairly easy to use the celery max retry functionality.